test/app-luatest/console_debugger_session_test.lua

```diff
@@ -298,25 +298,38 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
                 debuglog('Execute command: "'..trim(cmd)..'"')
             end
 
-            local result
-            local clean_cmd = trim(cmd)
-            -- there should be empty stderr - check it before stdout
-            local errout = fh:read({ timeout = 0.05, stderr = true})
-            debuglog('stderr output: ', trim(errout))
-            if first and errout then
-                -- we do not expect anything on stderr
-                -- with exception of initial debugger header
-                t.assert_str_contains(trim(errout), header, false)
+            if first then
+                -- only initial debugger header is expected on stderr
+                t.helpers.retrying({timeout = 3, delay = 0.05},
+                function(expected)
+                    local errout = trim(unescape(fh:read({ timeout = 0.5,
+                                                           stderr = true })))
+                    t.assert_str_contains(errout, expected, false)
+                    debuglog('stderr output: ', trim(errout))
+                end, header)
                 first = false
             else
+                -- stderr should be empty, check it before reading stdout
+                local errout = fh:read({ timeout = 0.05, stderr = true })
+                if errout == nil or trim(errout) == '' then
+                    debuglog('stderr output: ', trim(errout))
+                end
                 t.assert(errout == nil or trim(errout) == '')
             end
-            repeat
-                result = trim(unescape(fh:read({ timeout = 0.5 })))
-                debuglog('stdout output:', result)
-            until result ~= '' or result ~= clean_cmd
-            if expected ~= '' then
-                t.assert_str_contains(result, expected, false)
+
+            -- The output is returned in several chunks, first the prompt,
+            -- then the command, and finally the result. In most cases it
+            -- is can be read in one go, but to prevent flacking the reading
+            -- may be retried.
+            if (expected ~= '') then
+                t.helpers.retrying({timeout = 3, delay = 0.05},
+                function(expected)
+                    local result = trim(unescape(fh:read({ timeout = 0.5 })))
+                    t.assert_str_contains(result, expected, false)
+                    debuglog('stdout output:', result)
+                end, expected)
+            else
+                fh:read({ timeout = 0.5 })
```

Comment:
What is purpose of this `:read()`?

Reply:
To clear out the buffer for the followup reads. While we don't need to check the output in some rare cases, it seems like a good idea to separate it for different test cases by clearing the buffer.

Reply:
Answering me here doesn't make the idea obvious from the code. Please, address such review comments in the patchset itself.

However, I'm not satisfied by the answer. As for me, it is wrong in two ways:

1. If some output is produced by a test case that is expected to don't produce any output, the problem will be hidden (instead of failing the next test case). At least, I see no check that this `:read()` call is timed out.
2. `read()` system call (and, so, our `:read()` from the pipe) has no guarantee like 'it returns everything that were written on the other side'. If two bytes were written, we can receive just one. So, in fact, it is 'usually cleans the pipe buffer, but sometimes doesn't'. It is common for byte streams. Attempt to work with a byte stream like it is a message stream is wrong.

My proposals:

* Let the next test case fail. (And properly comment how it works.)
* Trigger some response that is guaranteed to be written after the response that is expected to be empty.

Reply:
I rewrote the test using updated `interactive_tarantool`.